src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h

```diff
@@ -1708,7 +1788,176 @@ namespace UiaOperationAbstraction
+        {
+            return *std::get<LocalType>(m_member);
+        }
+
+        const TupleLocalType& operator*() const
+        {
+            return *std::get<LocalType>(m_member);
+        }
+
+        UiaBool operator==(const UiaTuple& rhs) const
+        {
+            if (ShouldUseRemoteApi())
+            {
+                auto mutableThis = *this;
+                mutableThis.ToRemote();
+
+                auto mutableRhs = rhs;
+                mutableRhs.ToRemote();
+                return std::get<RemoteType>(mutableThis.m_member).IsEqual(std::get<RemoteType>(mutableRhs.m_member));
+            }
+
+            auto localTuple = std::get<LocalType>(m_member);
+            auto localRhsTuple = std::get<LocalType>(rhs.m_member);
+            return *localTuple == *localRhsTuple;
+        }
+
+        UiaBool operator!=(const UiaTuple& rhs) const
+        {
+            return !(*this == rhs);
+        }
+
+        void ToRemote()
+        {
+            auto delegator = UiaOperationScope::GetCurrentDelegator();
+            if (delegator)
+            {
+                delegator->ConvertVariantDataToRemote<ItemWrapperType...>(m_member);
+            }
+        }
+
+        template <unsigned int I>
+        void SetAt(TupleItemWrapperType<I> item)
+        {
+            if (ShouldUseRemoteApi())
+            {
+                UiaUint index{ I };
+                ToRemote();
+                index.ToRemote();
+                item.ToRemote();
+                std::get<RemoteType>(m_member).SetAt(index, item);
+                return;
+            }
+
+            std::get<I>(*std::get<LocalType>(m_member)) = item;
+        }
+
+        template <unsigned int I>
+        TupleItemWrapperType<I> GetAt()
+        {
+            if (ShouldUseRemoteApi())
+            {
+                UiaUint index{ I };
+                ToRemote();
+                index.ToRemote();
+                return TupleItemWrapperType<I>{ std::get<RemoteType>(m_member).GetAt(index) };
+            }
+
+            return std::get<I>(*std::get<LocalType>(m_member));
+        }
+
+        void FromRemoteResult(const winrt::Windows::Foundation::IInspectable& result)
+        {
+            auto operationResults = result.as<winrt::Windows::Foundation::Collections::IVector<winrt::Windows::Foundation::IInspectable>>();
+            m_member = std::make_shared<TupleLocalType>();
+
+            auto& localValue = std::get<LocalType>(m_member);
+            impl::RemoteArrayToLocalTuple<0, ItemWrapperType...>(operationResults, *localValue);
+        }
+
+    private:
+        template<std::size_t I, class ItemWrapperType, class... ItemWrapperTypes>
+        void SetTupleWrapperItems(ItemWrapperType&& item, ItemWrapperTypes&&... items)
+        {
+            SetAt<I>(std::forward<ItemWrapperType>(item));
+            SetTupleWrapperItems<I + 1>(std::forward<ItemWrapperTypes>(items)...);
+        }
+
+        template<std::size_t I>

```

Comment:
Is it worth using `if constexpr (sizeof...(ItemWrapperTypes) != 0)` to simplify this?

Reply:
I am not sure if that would simplify it... If I understand you correctly, you would want something like that:
```
        template<std::size_t I, class ItemWrapperType, class... ItemWrapperTypes>
        void SetTupleWrapperItems(ItemWrapperType&& item, ItemWrapperTypes&&... items)
        {
            SetAt<I>(std::forward<ItemWrapperType>(item));
            if constexpr (I < sizeof...(ItemWrapperTypes))
            {
                SetTupleWrapperItems<I + 1>(std::forward<ItemWrapperTypes>(items)...);
            }
        }
```
I am not comparing against `0`, because we need to have access to the current argument/item at least (to set it).

The reason why I went with the current approach was because `std::tuple<>` is a valid type if I remember correctly and therefore if we are mirroring `std::tuple`, I thought we should have kept it here as well.

Any thoughts around it? :)

Reply:
Do you mean this (`SetTupleWrapperItems`) can be called in two scenarios?
One is `SetTupleWrapperItems(UiaElement&& element, UiaBool&& value)`,
the other is `SetTupleWrapperItems() // no args at all`.
However, `UiaTuple() : UiaTuple(TupleLocalType{}) {}` is already defined. So does it mean `UiaTuple(ItemWrapperType... args)` will get at least one argument?
If not, when we call `UiaTuple t {};` which constructor will be called?

---
In reply to: [504948395](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/43#discussion_r504948395) [](ancestors = 504948395)

Reply:
The default constructor will. That's a good point. So technically, this would not really be a problem for the current constructor.

I am fine with going with either approach. So if you feel strongly against `0 !=`, I am happy to adjust it.

---
In reply to: [504968880](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/43#discussion_r504968880) [](ancestors = 504968880,504948395)

Reply:
Sure, not a big issue, let's keep the current.

Reply:
Fine by me. :)

---
In reply to: [505012556](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/43#discussion_r505012556) [](ancestors = 505012556)